rasa/shared/core/trackers.py

```diff
@@ -600,9 +623,15 @@ def update(self, event: Event, domain: Optional[Domain] = None) -> None:
         self.events.append(event)
         event.apply_to(self)
 
-        if domain and isinstance(event, UserUttered):
-            # store all entities as slots
-            for e in domain.slots_for_entities(event.parse_data["entities"]):
+        if domain and isinstance(event, (UserUttered, DefinePrevUserUtteredEntities)):
+            entities = event.entities
+            if isinstance(event, UserUttered):
+                # Rather get entities from `parse_data` as
+                # `DefinePrevUserUtteredEntities` might have already affected the
+                # `UserUttered.entities` attribute
+                entities = event.parse_data["entities"]
```

Comment:
could you please unpack this convoluted `if`, I think checks for `UserUttered` and `DefinePrevUserUtteredEntities` can be done in parallel

Reply:
actually, wouldn't `event.parse_data["entities"] == event.entities` for `DefinePrevUserUtteredEntities`?

also, I think we have a constant for `"entities"`, no?

Reply:
`# DefinePrevUserUtteredEntities might have already affected the UserUttered.entities attribute`? what does it mean? why behavior is not deterministic?

Reply:
> could you please unpack this convoluted if, I think checks for UserUttered and DefinePrevUserUtteredEntities can be done in parallel

I made it more explicit. 

> actually, wouldn't event.parse_data["entities"] == event.entities for DefinePrevUserUtteredEntities?

`DefinePrevUserUtteredEntities` doesn't have a `parse_data`?

> also, I think we have a constant for "entities", no?
 
done

Reply:
> DefinePrevUserUtteredEntities doesn't have a parse_data?

ok

what about `might` why it is not deterministic?

Reply:
I'm trying to play it save. Pretty sure that this would happen for  the `InMemoryTrackerStore`

Reply:
ok, could you please add comment, so it'll be easier to modify this piece in the future

Reply:
As you already merged, should I directly push the comment to `e2e` or are we good?

Reply:
sorry, forgot about this one. as it is just a comment could you please directly merge to `e2e`.

Reply:
done